tests/components/tado/test_coordinator.py

```diff
@@ -0,0 +1,70 @@
+"""Test the Tado coordinator."""
+
+from unittest.mock import patch
+
+import PyTado
+import PyTado.exceptions
+import pytest
+
+from homeassistant.components.tado.const import DOMAIN
+from homeassistant.components.tado.coordinator import TadoDataUpdateCoordinator
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError
+from homeassistant.helpers.update_coordinator import UpdateFailed
+
+from .util import async_init_integration
+
+from tests.common import MockConfigEntry, SnapshotAssertion
+
+
+@pytest.mark.parametrize(
+    ("exception", "expected_exception", "expected_message"),
+    [
+        (
+            PyTado.exceptions.TadoWrongCredentialsException("Invalid credentials"),
+            ConfigEntryError,
+            "Invalid Tado credentials. Error: Invalid credentials",
+        ),
+        (
+            PyTado.exceptions.TadoException("General error"),
+            UpdateFailed,
+            "Error during Tado setup: General error",
+        ),
+    ],
+)
+async def test_coordinator_setup_exceptions(
+    hass: HomeAssistant, exception, expected_exception, expected_message
+) -> None:
+    """Test the Tado coordinator setup exceptions."""
+
+    config_entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={"username": "mock", "password": "mock"},
+        options={"fallback": "NEXT_TIME_BLOCK"},
+    )
+    config_entry.add_to_hass(hass)
+
+    coordinator = TadoDataUpdateCoordinator(
+        hass=hass, username="mock", password="mock", fallback="NEXT_TIME_BLOCK"
+    )
```

Comment:
So we should not create our own coordinator in the tests. Instead we should test more end-to-end. So in this test you're testing setup exceptions. Rather mock the library, have HA setup the integration and then observe the config entry state